pkg/controller/podautoscaler/metrics/metrics_client.go

```diff
@@ -156,14 +160,24 @@ func (h *HeapsterMetricsClient) getCpuUtilizationForPods(namespace string, selec
 	metricPath := fmt.Sprintf("/apis/metrics/v1alpha1/namespaces/%s/pods", namespace)
 	params := map[string]string{"labelSelector": selector.String()}
 
-	resultRaw, err := h.client.Core().Services(h.heapsterNamespace).
-		ProxyGet(h.heapsterScheme, h.heapsterService, h.heapsterPort, metricPath, params).
-		DoRaw()
+	var (
+		resultRaw []byte
+		err       error
+	)
+
+	if h.heapsterUrl != "" {
```

Comment:
You have duplicated logic here and in line 254. How about having a method queryHeapster which will choose appropriate way to query heapster (using service discovery or external address) and return (resultRaw, err) pair?